### PR TITLE
Remove ioredis; upstream has types

### PR DESCRIPTION
**Pull request by jakebailey**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If removing a declaration:
- [x] ~If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)~
- [x] Delete the package's directory.
- [x] Add it to `notNeededPackages.json`.

### CODE REVIEW

File: types/ioredis-mock/ioredis-mock-tests.ts [hunk 1]
```diff
@@ -69,8 +69,8 @@ const f = async () => {
     // now you can see we have one new message
 
     // use xread to read all messages in channel
-    const messages1 = await redis.xread(['STREAMS', channel, 0]);
-    const messages2 = messages1[0][1];
+    const messages1 = await redis.xread('STREAMS', channel, 0);
```
Comment (jakebailey) [inline types/ioredis-mock/ioredis-mock-tests.ts:72]: Confusingly, the ioredis examples don't compile using their own types: https://github.com/luin/ioredis/blob/308017a6b9429c16b074e03e70f5524499476fa9/examples/redis_streams.js#L28

File: types/ioredis-mock/index.d.ts [hunk 2]
```diff
@@ -1,42 +1,8 @@
...
-export type Pipeline = RealIORedis.Pipeline;
-export type PreferredSlaves = RealIORedis.PreferredSlaves;
-export type Redis = RealIORedis.Redis;
-export type RedisOptions = RealIORedis.RedisOptions;
-export type ScanStreamOption = RealIORedis.ScanStreamOption;
-export type SecureContextOptions = RealIORedis.SecureContextOptions;
-export type SecureVersion = RealIORedis.SecureVersion;
-export type ValueType = RealIORedis.ValueType;
+import ioredis = require('ioredis');
+export = ioredis;
```
Comment (jakebailey) [inline types/ioredis-mock/index.d.ts:8]: I _think_ this is the right way to reexport the other package.

**Comment by 43081j** [comment]
coincidentally was just about to do the same PR. its a hefty one :D

there's some dependent types that need updating to fix your build:

* bull-arena (no clue why this actually uses ioredis, the source package doesn't depend on it directly)
* koa-ratelimit
* rate-limit-redis - **ships its own types too now** ([here](https://unpkg.com/browse/rate-limit-redis@3.0.1/dist/index.d.ts))
* ratelimiter
* redlock

**Comment by jakebailey** [comment]
Yep, currently working on it, though doing other stuff at the same time.

I sent #62885 for rate-limit-redis already, but apparently there are no reviewers. I'll probably just merge it.

File: types/ratelimiter/ratelimiter-tests.ts [hunk 3]
```diff
@@ -1,19 +1,16 @@
-import * as redis from 'redis';
-import * as ioredis from 'ioredis';
 import Limiter = require('ratelimiter');
 
 declare let id: string;
-declare let db: redis.RedisClient;
-declare let iodb: ioredis.Redis;
+declare let db: {
+    multi(operations: any[][]): { exec(cb: (err: any, res: any) => unknown): void };
+};
```
Comment (jakebailey) [inline types/ratelimiter/ratelimiter-tests.ts:6]: Generally speaking, it's not so great to have dependencies just so they are used in tests, hence this removal.

File: types/bull-arena/package.json [hunk 4]
```diff
@@ -1,6 +1,7 @@
 {
     "private": true,
     "dependencies": {
+        "@types/ioredis": "^4.28.10",
```
Comment (jakebailey) [inline types/bull-arena/package.json:4]: bull-area appears to be unmaintained (per a pinned issue on their repo); I'm not going to attempt to figure out how to make this work with ioredis 5.

File: types/koa-ratelimit/index.d.ts [hunk 5]
```diff
@@ -38,7 +37,7 @@ declare namespace KoaRatelimit {
         /**
          * The database powering the backing rate-limiter package.
          */
-        db: Redis | RedisClient | Map<any, any>;
+        db: Redis | Map<any, any>;
```
Comment (jakebailey) [inline types/koa-ratelimit/index.d.ts:40]: It turns out that kop-ratelimit is just a wrapper on async-ratelimit: https://github.com/koajs/ratelimit/blob/master/limiter/redis.js

async-ratelimit is typed, and does not support `redis`, only `ioredis`.

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## ioredis-mock ([<kbd>unpkg</kbd>](https://unpkg.com/browse/ioredis-mock@latest/))
was missing the following properties:
1. The declaration doesn't match the JavaScript module 'ioredis-mock'. Reason:
The JavaScript module can be called or constructed, but the declaration module cannot.
1. once
1. on
1. listenerCount
1. errorMonitor

<details>
<summary>as well as these 4 other properties...</summary>
<p>captureRejectionSymbol, captureRejections, defaultMaxListeners, EventEmitter</p>
</details>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against cc13fe371c105348730e8d0dd47760b66ea58aff
</p>

**Review by plantain-00 — APPROVED** [review]

File: types/bull-arena/package.json (hunk 4 above)
Comment (peterblazejewicz) [reply types/bull-arena/package.json:4]: this is perfectly fine approach here IMO

**Review by peterblazejewicz — APPROVED** [review]
LGTM!

**Review by lenovouser — APPROVED** [review]

**Pull request merged**